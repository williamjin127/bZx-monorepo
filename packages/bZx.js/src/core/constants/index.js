export const SOLIDITY_TYPE_MAX_CHARS = 64;
