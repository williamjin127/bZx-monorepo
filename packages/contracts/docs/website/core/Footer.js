/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const React = require("react");

const siteConfig = require(process.cwd() + "/siteConfig.js");

class Footer extends React.Component {
  docUrl(doc, language) {
    const baseUrl = this.props.config.baseUrl;
    return baseUrl + "docs/" + (language ? language + "/" : "") + doc;
  }

  pageUrl(doc, language) {
    const baseUrl = this.props.config.baseUrl;
    return baseUrl + (language ? language + "/" : "") + doc;
  }

  render() {
    const currentYear = new Date().getFullYear();
    return (
      <footer className="nav-footer" id="footer">
        <section className="sitemap">
          <a href={this.props.config.baseUrl} className="nav-home">
            {this.props.config.footerIcon && (
              <img
                src={this.props.config.baseUrl + this.props.config.footerIcon}
                alt={this.props.config.title}
                width="66"
                height="58"
              />
            )}
          </a>
          <div>
            <h5>Docs</h5>
            <a href={this.docUrl("migrations_Migrations.html")}>
              Getting Started
            </a>
          </div>
          <div>
            <h5>Community</h5>
            <a
              href={siteConfig.stackoverflow}
              target="_blank"
              rel="noreferrer noopener"
            >
              Stack Overflow
            </a>
            <a
              href={siteConfig.twitter}
              target="_blank"
              rel="noreferrer noopener"
            >
              Twitter
            </a>
          </div>
          <div>
            <h5>More</h5>
            <a href={siteConfig.medium} target="_blank">
              Medium
            </a>
            <a href={siteConfig.githubCompanyRepository} target="_blank">
              GitHub
            </a>
            <a
              className="github-button"
              href={siteConfig.githubRepository}
              data-icon="octicon-star"
              data-count-href="/bZxNetwork/protocol_contracts/stargazers"
              data-show-count={true}
              data-count-aria-label="# stargazers on GitHub"
              aria-label="Star this project on GitHub"
            >
              Star
            </a>
          </div>
        </section>
        <section className="copyright">{this.props.config.copyright}</section>
      </footer>
    );
  }
}

module.exports = Footer;
