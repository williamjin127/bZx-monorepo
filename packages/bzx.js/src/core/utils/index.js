import { BigNumber } from "@0xproject/utils";
import { assert } from "@0xproject/assert";
import BN from "bn.js";
import Web3Utils from "web3-utils";
import * as constants from "../constants";
import { SchemaValidator } from "../../schemas/bZx_json_schemas";
import { getContracts } from "../../contracts";
import * as Addresses from "../../addresses";

export const noop = () => {};

export const bigNumberToBN = value => new BN(value.toString(), 10);

export const generatePseudoRandomSalt = () => {
  // BigNumber.random returns a pseudo-random number between 0 & 1
  // with a passed in number of decimal places.
  // Source: https://mikemcl.github.io/bignumber.js/#random
  const randomNumber = BigNumber.random(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT
  );
  const factor = new BigNumber(10).pow(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT - 1
  );
  const salt = randomNumber.times(factor).round();
  return salt;
};

const getOrderValues = (order, shouldFormatAsStrings) => {
  // Must be strings in production for Web3Utils.soliditySha3 for some reason
  if (shouldFormatAsStrings) {
    return [
      order.loanTokenAmount.toString(),
      order.interestAmount.toString(),
      order.initialMarginAmount.toString(),
      order.maintenanceMarginAmount.toString(),
      order.lenderRelayFee.toString(),
      order.traderRelayFee.toString(),
      order.maxDurationUnixTimestampSec.toString(),
      order.expirationUnixTimestampSec.toString(),
      order.makerRole.toString(),
      order.withdrawOnOpen.toString(),
      order.salt.toString()
    ];
  }
  return [
    bigNumberToBN(order.loanTokenAmount),
    bigNumberToBN(order.interestAmount),
    bigNumberToBN(order.initialMarginAmount),
    bigNumberToBN(order.maintenanceMarginAmount),
    bigNumberToBN(order.lenderRelayFee),
    bigNumberToBN(order.traderRelayFee),
    bigNumberToBN(order.maxDurationUnixTimestampSec),
    bigNumberToBN(order.expirationUnixTimestampSec),
    bigNumberToBN(order.makerRole),
    bigNumberToBN(order.withdrawOnOpen),
    bigNumberToBN(order.salt)
  ];
};

const getLoanOrderHashArgs = (order, shouldFormatAsStrings) => {
  const orderAddresses = [
    order.makerAddress,
    order.loanTokenAddress,
    order.interestTokenAddress,
    order.collateralTokenAddress,
    order.feeRecipientAddress,
    order.oracleAddress,
    order.takerAddress,
    order.tradeTokenToFillAddress
  ];
  const orderValues = getOrderValues(order, shouldFormatAsStrings);
  const oracleData = order.oracleData || "0x";

  return { orderAddresses, orderValues, oracleData };
};

export const doesContractExistAtAddress = async (web3, address) => {
  const code = await web3.eth.getCode(address);
  // Regex matches 0x0, 0x00, 0x in order to accommodate poorly implemented clients
  const codeIsEmpty = /^0x0{0,40}$/i.test(code);
  return !codeIsEmpty;
};

export const getContractInstance = (web3, abi, address) => {
  assert.isETHAddressHex("address", address);
  const contract = new web3.eth.Contract(abi, address);
  return contract;
};

export const getLoanOrderHashHex = order => {
  const { orderAddresses, orderValues, oracleData } = getLoanOrderHashArgs(order, true);
  const orderHashHex = Web3Utils.soliditySha3(
    { t: "address", v: order.bZxAddress },
    { t: "address[8]", v: orderAddresses },
    { t: "uint256[11]", v: orderValues },
    { t: "bytes", v: oracleData }
  );
  return orderHashHex;
};

export const getLoanOrderHashAsync = async ({ web3, networkId }, order) => {
  const { orderAddresses, orderValues, oracleData } = getLoanOrderHashArgs(order, true);
  const bZxContract = await getContractInstance(
    web3,
    getContracts(networkId).BZx.abi,
    Addresses.getAddresses(networkId).BZx
  );
  return bZxContract.methods
    .getLoanOrderHash(
      orderAddresses,
      orderValues,
      oracleData
    )
    .call();
};

export const doesConformToSchema = (variableName, value, schema) => {
  const schemaValidator = new SchemaValidator();
  const validationResult = schemaValidator.validate(value, schema);
  const hasValidationErrors = validationResult.errors.length > 0;
  const msg = `Expected ${variableName} to conform to schema ${
    schema.id
  }\nEncountered: ${JSON.stringify(
    value,
    null,
    "\t"
  )}\nValidation errors: ${validationResult.errors.join(", ")}`;
  assert.assert(!hasValidationErrors, msg);
};

export const toChecksumAddress = addr => Web3Utils.toChecksumAddress(addr);

export const requestFaucetToken = (
  { web3, networkId },
  { tokenAddress, receiverAddress, getObject, txOpts }
) => {
  const faucetContract = getContractInstance(
    web3,
    getContracts(networkId).TestNetFaucet.abi,
    Addresses.getAddresses(networkId).TestNetFaucet
  );

  const txObj = faucetContract.methods.faucet(
    toChecksumAddress(tokenAddress),
    toChecksumAddress(receiverAddress)
  );
  console.log(`requestFaucetToken: ${txObj.encodeABI()}`);

  if (getObject) {
    return txObj;
  }
  return txObj.send(txOpts);
};

export const getWeb3Contract = ({ web3, networkId }, contractName, contractAddress) => getContractInstance(
    web3,
    getContracts(networkId)[contractName].abi,
    contractAddress ? contractAddress : Addresses.getAddresses(networkId)[contractName]
  );
