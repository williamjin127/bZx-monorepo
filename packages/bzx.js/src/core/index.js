import { assert } from "@0xproject/assert";
import { BigNumber } from "@0xproject/utils";
import * as constants from "./constants";
import { schemas } from "../schemas/bZx_json_schemas";
import * as utils from "./utils";
import * as tokenRegistry from "../tokenRegistry";
import EIP20 from "../contracts/EIP20.json";
import * as allowance from "../allowance";
import * as oracles from "../oracles";
import * as fill from "../fill";
import * as Addresses from "../addresses";
import * as orderHistory from "../orderHistory";
import * as transfer from "../transfer";
import * as signature from "../signature";
import * as Errors from "./constants/errors";
import * as trade from "../trade";
import * as loanHealth from "../loanHealth";
import * as bounty from "../bounty";
import * as weth from "../weth";

export class BZxJS {
  static generatePseudoRandomSalt = utils.generatePseudoRandomSalt;
  static noop = utils.noop;
  static toChecksumAddress = utils.toChecksumAddress;

  /* On Metamask, provider.host is undefined
  Force users to provide host url */
  constructor(web3, { networkId, addresses = Addresses.getAddresses(networkId) } = {}) {
    if (!networkId) throw new Error(Errors.NoNetworkId);

    this.web3 = web3;
    this.addresses = addresses;
    this.networkId = networkId;
    switch (networkId) {
      case 1:
        this.networkName = "mainnet";
        this.etherscanURL = "https://etherscan.io/";
        break;
      case 3:
        this.networkName = "ropsten";
        this.etherscanURL = "https://ropsten.etherscan.io/";
        break;
      case 4:
        this.networkName = "rinkeby";
        this.etherscanURL = "https://rinkeby.etherscan.io/";
        break;
      case 42:
        this.networkName = "kovan";
        this.etherscanURL = "https://kovan.etherscan.io/";
        break;
      default:
        this.networkName = "local";
        this.etherscanURL = "";
        break;
    }
  }

  static getLoanOrderHashHex(order) {
    utils.doesConformToSchema("loanOrder", order, schemas.loanOrderSchema);
    const orderHashHex = utils.getLoanOrderHashHex(order);
    return orderHashHex;
  }
  getLoanOrderHashAsync = async props =>
    utils.getLoanOrderHashAsync(this, props);

  static isValidSignature = props => signature.isValidSignature(props);

  isValidSignatureAsync = async props =>
    signature.isValidSignatureAsync(this, props);

  signOrderHashAsync = async (...props) =>
    signature.signOrderHashAsync(this, ...props);

  setAllowance = (...props) => allowance.setAllowance(this, ...props);

  setAllowanceUnlimited = props =>
    this.setAllowance({
      ...props,
      amountInBaseUnits: constants.UNLIMITED_ALLOWANCE_IN_BASE_UNITS
    });

  resetAllowance = props =>
    this.setAllowance({
      ...props,
      amountInBaseUnits: new BigNumber(0)
    });

  getAllowance = async (...props) => allowance.getAllowance(this, ...props);

  getBalance = async ({ tokenAddress, ownerAddress }) => {
    assert.isETHAddressHex("ownerAddress", ownerAddress);
    assert.isETHAddressHex("tokenAddress", tokenAddress);

    const tokenContract = await utils.getContractInstance(
      this.web3,
      EIP20.abi,
      tokenAddress
    );
    const balance = await tokenContract.methods.balanceOf(ownerAddress).call();
    return new BigNumber(balance);
  };

  getTokenList = async () => tokenRegistry.getTokenList(this);

  getOracleList = async () => oracles.getOracleList(this);
  isTradeSupported = async (...props) =>
    oracles.isTradeSupported(this, ...props);
  getConversionData = async (...props) =>
    oracles.getConversionData(this, ...props);

  takeLoanOrderAsLender = (...props) =>
    fill.takeLoanOrderAsLender(this, ...props);

  takeLoanOrderAsTrader = (...props) =>
    fill.takeLoanOrderAsTrader(this, ...props);

  pushLoanOrderOnChain = (...props) =>
    fill.pushLoanOrderOnChain(this, ...props);

  takeLoanOrderOnChainAsTrader = (...props) =>
    fill.takeLoanOrderOnChainAsTrader(this, ...props);

  takeLoanOrderOnChainAsLender = (...props) =>
    fill.takeLoanOrderOnChainAsLender(this, ...props);

  cancelLoanOrder = (...props) => fill.cancelLoanOrder(this, ...props);

  cancelLoanOrderWithHash = (...props) =>
    fill.cancelLoanOrderWithHash(this, ...props);

  getInitialCollateralRequired = async (...props) =>
    fill.getInitialCollateralRequired(this, ...props);

  orderFilledAmount = async (...props) =>
    fill.orderFilledAmount(this, ...props);
  orderCancelledAmount = async (...props) =>
    fill.orderCancelledAmount(this, ...props);

  getSingleOrder = async (...props) =>
    orderHistory.getSingleOrder(this, ...props);
  getOrdersFillable = async (...props) =>
    orderHistory.getOrdersFillable(this, ...props);
  getOrdersForUser = async (...props) =>
    orderHistory.getOrdersForUser(this, ...props);
  getSingleLoan = async (...props) =>
    orderHistory.getSingleLoan(this, ...props);
  getLoansForLender = async (...props) =>
    orderHistory.getLoansForLender(this, ...props);
  getLoansForTrader = async (...props) =>
    orderHistory.getLoansForTrader(this, ...props);

  transferToken = (...props) => transfer.transferToken(this, ...props);

  tradePositionWith0x = (...props) => trade.tradePositionWith0x(this, ...props);
  tradePositionWith0xV2 = (...props) => trade.tradePositionWith0xV2(this, ...props);
  tradePositionWithOracle = (...props) =>
    trade.tradePositionWithOracle(this, ...props);

  depositCollateral = (...props) => loanHealth.depositCollateral(this, ...props);
  withdrawCollateral = (...props) => loanHealth.withdrawCollateral(this, ...props);
  changeCollateral = (...props) => loanHealth.changeCollateral(this, ...props);
  withdrawPosition = (...props) => loanHealth.withdrawPosition(this, ...props);
  depositPosition = (...props) => loanHealth.depositPosition(this, ...props);
  getPositionOffset = (...props) => loanHealth.getPositionOffset(this, ...props);

  closeLoanPartially = (...props) => loanHealth.closeLoanPartially(this, ...props);
  closeLoan = (...props) => loanHealth.closeLoan(this, ...props);

  payInterestForOracle = (...props) => loanHealth.payInterestForOracle(this, ...props);
  getLenderInterestForOracle = (...props) => loanHealth.getLenderInterestForOracle(this, ...props);
  getTraderInterestForLoan = (...props) => loanHealth.getTraderInterestForLoan(this, ...props);

  requestFaucetToken = (...props) => utils.requestFaucetToken(this, ...props);

  getActiveLoans = (...props) => bounty.getActiveLoans(this, ...props);
  getMarginLevels = (...props) => bounty.getMarginLevels(this, ...props);
  liquidateLoan = (...props) => bounty.liquidateLoan(this, ...props);

  wrapEth = (...props) => weth.wrapEth(this, ...props);
  unwrapEth = (...props) => weth.unwrapEth(this, ...props);

  getWeb3Contract = (...props) => utils.getWeb3Contract(this, ...props);
}

export default BZxJS;
