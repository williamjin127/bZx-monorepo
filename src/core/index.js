import { assert } from "@0xproject/assert";
import { constants } from "0x.js/lib/src/utils/constants";
import { BigNumber } from "@0xproject/utils";
import { schemas } from "../schemas/bZx_json_schemas";
import * as utils from "./utils";
import * as tokenRegistry from "../tokenRegistry";
import EIP20 from "../contracts/EIP20.json";
import * as allowance from "../allowance";
import * as oracles from "../oracles";
import * as fill from "../fill";
import * as Addresses from "../addresses";
import * as orderHistory from "../orderHistory";
import * as transfer from "../transfer";
import * as signature from "../signature";
import * as Errors from "./constants/errors";
import * as trade from "../trade";
import * as loanHealth from "../loanHealth";
import * as bounty from "../bounty";
import * as weth from "../weth";

const Web3 = require("web3"); // eslint-disable global-require

export default class BZxJS {
  static generatePseudoRandomSalt = utils.generatePseudoRandomSalt;
  static noop = utils.noop;
  static toChecksumAddress = utils.toChecksumAddress;

  /* On Metamask, provider.host is undefined
  Force users to provide host url */
  constructor(
    provider,
    { networkId, addresses = Addresses.getAddresses(networkId) } = {}
  ) {
    if (!networkId) throw new Error(Errors.NoNetworkId);

    assert.isWeb3Provider("provider", provider);
    this.web3 = new Web3(provider);
    this.addresses = addresses;
    this.networkId = networkId;
    switch (networkId) {
      case 1:
        this.networkName = "mainnet";
        this.etherscanURL = "https://etherscan.io/";
        break;
      case 3:
        this.networkName = "ropsten";
        this.etherscanURL = "https://ropsten.etherscan.io/";
        break;
      case 4:
        this.networkName = "rinkeby";
        this.etherscanURL = "https://rinkeby.etherscan.io/";
        break;
      case 42:
        this.networkName = "kovan";
        this.etherscanURL = "https://kovan.etherscan.io/";
        break;
      default:
        this.networkName = "local";
        this.etherscanURL = "";
        break;
    }
  }

  static getLoanOrderHashHex(order) {
    utils.doesConformToSchema("loanOrder", order, schemas.loanOrderSchema);
    const orderHashHex = utils.getLoanOrderHashHex(order);
    return orderHashHex;
  }
  getLoanOrderHashAsync = async props =>
    utils.getLoanOrderHashAsync(this, props);

  static isValidSignature = props => signature.isValidSignature(props);

  isValidSignatureAsync = async props =>
    signature.isValidSignatureAsync(this, props);

  signOrderHashAsync = async (...props) =>
    signature.signOrderHashAsync(this, ...props);

  setAllowance = (...props) => allowance.setAllowance(this, ...props);

  setAllowanceUnlimited = props =>
    this.setAllowance({
      ...props,
      amountInBaseUnits: constants.UNLIMITED_ALLOWANCE_IN_BASE_UNITS
    });

  resetAllowance = props =>
    this.setAllowance({
      ...props,
      amountInBaseUnits: new BigNumber(0)
    });

  getAllowance = async (...props) => allowance.getAllowance(this, ...props);

  getBalance = async ({ tokenAddress, ownerAddress }) => {
    assert.isETHAddressHex("ownerAddress", ownerAddress);
    assert.isETHAddressHex("tokenAddress", tokenAddress);

    const tokenContract = await utils.getContractInstance(
      this.web3,
      EIP20.abi,
      tokenAddress
    );
    const balance = await tokenContract.methods.balanceOf(ownerAddress).call();
    return new BigNumber(balance);
  };

  getTokenList = async () => tokenRegistry.getTokenList(this);

  getOracleList = async () => oracles.getOracleList(this);
  isTradeSupported = async (...props) =>
    oracles.isTradeSupported(this, ...props);

  takeLoanOrderAsLender = (...props) =>
    fill.takeLoanOrderAsLender(this, ...props);

  takeLoanOrderAsTrader = (...props) =>
    fill.takeLoanOrderAsTrader(this, ...props);

  getInitialCollateralRequired = async (...props) =>
    fill.getInitialCollateralRequired(this, ...props);

  getSingleOrder = async (...props) =>
    orderHistory.getSingleOrder(this, ...props);
  getOrders = async (...props) =>
    orderHistory.getOrders(this, ...props);
  getSingleLoan = async (...props) =>
    orderHistory.getSingleLoan(this, ...props);
  getLoansForLender = async (...props) =>
    orderHistory.getLoansForLender(this, ...props);
  getLoansForTrader = async (...props) =>
    orderHistory.getLoansForTrader(this, ...props);

  transferToken = (...props) => transfer.transferToken(this, ...props);

  tradePositionWith0x = (...props) => trade.tradePositionWith0x(this, ...props);
  tradePositionWithOracle = (...props) =>
    trade.tradePositionWithOracle(this, ...props);

  changeCollateral = (...props) => loanHealth.changeCollateral(this, ...props);
  depositCollateral = (...props) =>
    loanHealth.depositCollateral(this, ...props);
  withdrawExcessCollateral = (...props) =>
    loanHealth.withdrawExcessCollateral(this, ...props);
  getProfitOrLoss = (...props) => loanHealth.getProfitOrLoss(this, ...props);
  withdrawProfit = (...props) => loanHealth.withdrawProfit(this, ...props);

  closeLoan = (...props) => loanHealth.closeLoan(this, ...props);
  payInterest = (...props) => loanHealth.payInterest(this, ...props);

  requestFaucetToken = (...props) => utils.requestFaucetToken(this, ...props);

  getActiveLoans = (...props) => bounty.getActiveLoans(this, ...props);
  getMarginLevels = (...props) => bounty.getMarginLevels(this, ...props);
  liquidateLoan = (...props) => bounty.liquidateLoan(this, ...props);

  wrapEth = (...props) => weth.wrapEth(this, ...props);
  unwrapEth = (...props) => weth.unwrapEth(this, ...props);
}
