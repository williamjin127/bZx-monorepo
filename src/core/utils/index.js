import { BigNumber } from "@0xproject/utils";
import { assert } from "@0xproject/assert";
import { constants } from "0x.js/lib/src/utils/constants";
import BN from "bn.js";
import Web3Utils from "web3-utils";
import * as Errors from "../constants/errors";
import { SchemaValidator } from "../../schemas/b0x_json_schemas";
import { ropsten as Contracts } from "../../contracts";
import * as Addresses from "../../addresses";

export const noop = () => {};

export const bigNumberToBN = value => new BN(value.toString(), 10);

export const generatePseudoRandomSalt = () => {
  // BigNumber.random returns a pseudo-random number between 0 & 1
  // with a passed in number of decimal places.
  // Source: https://mikemcl.github.io/bignumber.js/#random
  const randomNumber = BigNumber.random(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT
  );
  const factor = new BigNumber(10).pow(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT - 1
  );
  const salt = randomNumber.times(factor).round();
  return salt;
};

export const getLoanOrderHashHex = order => {
  const orderAddrs = [
    order.makerAddress,
    order.loanTokenAddress,
    order.interestTokenAddress,
    order.collateralTokenAddress,
    order.feeRecipientAddress,
    order.oracleAddress
  ];
  const orderUints = [
    bigNumberToBN(order.loanTokenAmount),
    bigNumberToBN(order.interestAmount),
    bigNumberToBN(order.initialMarginAmount),
    bigNumberToBN(order.maintenanceMarginAmount),
    bigNumberToBN(order.lenderRelayFee),
    bigNumberToBN(order.traderRelayFee),
    bigNumberToBN(order.expirationUnixTimestampSec),
    bigNumberToBN(order.makerRole),
    bigNumberToBN(order.salt)
  ];
  const orderHashHex = Web3Utils.soliditySha3(
    { t: "address", v: order.b0xAddress },
    { t: "address[6]", v: orderAddrs },
    { t: "uint256[9]", v: orderUints }
  );
  return orderHashHex;
};

export const doesContractExistAtAddress = async (web3, address) => {
  const code = await web3.eth.getCode(address);
  // Regex matches 0x0, 0x00, 0x in order to accommodate poorly implemented clients
  const codeIsEmpty = /^0x0{0,40}$/i.test(code);
  return !codeIsEmpty;
};

export const getContractInstance = async (web3, abi, address) => {
  assert.isETHAddressHex("address", address);
  const contractExists = await doesContractExistAtAddress(web3, address);
  if (!contractExists)
    throw new Error(`${Errors.ContractDoesNotExist} ${address}`);

  const contract = new web3.eth.Contract(abi, address);
  return contract;
};

export const doesConformToSchema = (variableName, value, schema) => {
  const schemaValidator = new SchemaValidator();
  const validationResult = schemaValidator.validate(value, schema);
  const hasValidationErrors = validationResult.errors.length > 0;
  const msg = `Expected ${variableName} to conform to schema ${
    schema.id
  }\nEncountered: ${JSON.stringify(
    value,
    null,
    "\t"
  )}\nValidation errors: ${validationResult.errors.join(", ")}`;
  assert.assert(!hasValidationErrors, msg);
};

export const isValidSignature = async (
  web3,
  { account, orderHash, signature }
) => {
  const b0xContract = await getContractInstance(
    web3,
    Contracts.B0x.abi,
    Addresses.getAddresses(web3.currentProvider).B0x
  );
  return b0xContract.methods
    .isValidSignature(account, orderHash, signature)
    .call();
};
