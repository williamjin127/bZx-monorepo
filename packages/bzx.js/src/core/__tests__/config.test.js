import Web3 from "web3";
import { BZxJS } from "../../core";

describe("config", () => {
  test("should use default config when config arg not provided", () => {
    const networkUrl = "http://localhost:8545";
    const provider = new Web3.providers.HttpProvider(networkUrl);
    const bZxJS = new BZxJS(provider, { networkId: 50 });

    expect(bZxJS.addresses).toMatchSnapshot();
  });

  test("should use provided config", () => {
    const networkUrl = "http://localhost:8545";
    const provider = new Web3.providers.HttpProvider(networkUrl);
    const config = {
      networkId: 50,
      addresses: {
        BZx: "thisisatest",
        BZxOracle: "thisisatest",
        BZxTo0x: "thisisatest",
        BZxToken: "thisisatest",
        BZxVault: "thisisatest",
        EIP20: "thisisatest",
        OracleRegistry: "thisisatest",
        OracleInterface: "thisisatest",
        TestToken0: "thisisatest",
        TestToken1: "thisisatest",
        TestToken2: "thisisatest",
        TestToken3: "thisisatest",
        TestToken4: "thisisatest",
        TestToken5: "thisisatest",
        TestToken6: "thisisatest",
        TestToken7: "thisisatest",
        TestToken8: "thisisatest",
        TestToken9: "thisisatest",
        TokenRegistry: "thisisatest"
      }
    };

    const bZxJS = new BZxJS(provider, config);

    expect(bZxJS.addresses).toMatchSnapshot();
  });
});
