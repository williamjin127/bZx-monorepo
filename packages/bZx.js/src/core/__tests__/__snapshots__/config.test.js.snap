// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config should use default config when config arg not provided 1`] = `
Object {
  "BZx": "0x8726c7414ac023d23348326b47af3205185fd035",
  "BZxOracle": "0x0d8b0dd11f5d34ed41d556def5f841900d5b1c6b",
  "BZxTo0x": "0x32eecaf51dfea9618e9bc94e9fbfddb1bbdcba15",
  "BZxToken": "0x04b5dadd2c0d6a261bfafbc964e0cac48585def3",
  "BZxVault": "0xb7c9b454221e26880eb9c3101b3295ca7d8279ef",
  "EIP20": "",
  "OracleInterface": "",
  "OracleRegistry": "0x6000eca38b8b5bba64986182fe2a69c57f6b5414",
  "TestNetFaucet": "0x609acc8b356894a937fc58f3411f9528de96ecb1",
  "TestToken0": "0xb48e1b16829c7f5bd62b76cb878a6bb1c4625d7a",
  "TestToken1": "0xc4cc602a7345518d0b7a84049d4bc8575ebf3398",
  "TestToken2": "0xe704967449b57b2382b7fa482718748c13c63190",
  "TestToken3": "0xa4b3e1659c473623287b2cc13b194705cd792525",
  "TestToken4": "0x2c530e4ecc573f11bd72cf5fdf580d134d25f15f",
  "TestToken5": "0x72d5a2213bfe46df9fbda08e22f536ac6ca8907e",
  "TestToken6": "0x2ebb94cc79d7d0f1195300aaf191d118f53292a8",
  "TestToken7": "0x5315e44798395d4a952530d131249fe00f554565",
  "TestToken8": "0xdff540fe764855d3175dcfae9d91ae8aee5c6d6f",
  "TestToken9": "0xc1be2c0bb387aa13d5019a9c518e8bc93cb53360",
  "TokenRegistry": "0x8a063452f7df2614db1bca3a85ef35da40cf0835",
}
`;

exports[`config should use provided config 1`] = `
Object {
  "BZx": "thisisatest",
  "BZxOracle": "thisisatest",
  "BZxTo0x": "thisisatest",
  "BZxToken": "thisisatest",
  "BZxVault": "thisisatest",
  "EIP20": "thisisatest",
  "OracleInterface": "thisisatest",
  "OracleRegistry": "thisisatest",
  "TestToken0": "thisisatest",
  "TestToken1": "thisisatest",
  "TestToken2": "thisisatest",
  "TestToken3": "thisisatest",
  "TestToken4": "thisisatest",
  "TestToken5": "thisisatest",
  "TestToken6": "thisisatest",
  "TestToken7": "thisisatest",
  "TestToken8": "thisisatest",
  "TestToken9": "thisisatest",
  "TokenRegistry": "thisisatest",
}
`;
